snap/snapcraft.yaml

```diff
@@ -24,43 +25,125 @@ layout:
+      # built with a simple `setup.py install`. Building a `bdist_wheel` and
+      # then installing it via `pip` addresses this shortcoming.
+      ${SNAPCRAFT_PART_INSTALL}/bin/python3 setup.py bdist_wheel
+      ${SNAPCRAFT_PART_INSTALL}/bin/pip install dist/hotsos-*.whl
+
+      # Update repo-info
+      git rev-parse --short HEAD > repo-info
+      # Install repo-info
+      install --mode 0644 repo-info ${SNAPCRAFT_PART_INSTALL}
+
+      # Rewrite the shebang to point to the correct Python interpreter. Only
+      # touch the first line and absorb the whole line. Note that the matching
+      # expression cannot be more specific since the path to the Python
+      # interpreter depends on the build environment. Notably, local builds
+      # (`/root`) differ from builds on launchpad (`/build`).
+      sed --in-place \
+        --expression "1 s:^#\!/.*$:#\!/snap/hotsos/current/usr/bin/${PYTHON}:" \
+        ${SNAPCRAFT_PART_INSTALL}/bin/hotsos
+
+      # Add a missing link in snap to Python interpreter.
+      ln -sf ../usr/bin/${PYTHON} ${SNAPCRAFT_PART_INSTALL}/bin/python3
+
+      # Do not use `craftctl default` and `organize` because this simply copies
+      # the whole repository to the install directory. Using rsync on `defs` is
+      # cleaner.
+      mkdir -p ${SNAPCRAFT_PART_INSTALL}/etc/hotsos
+      rsync -av defs ${SNAPCRAFT_PART_INSTALL}/etc/hotsos/
+    stage:
       # need to exclude these files since they are dangling symlinks
       # that cause the snapstore to reject the upload.
-      exclude-systemd-files:
-        - -lib/systemd/system/hwclock.service
-        - -lib/systemd/system/rc.service
-        - -lib/systemd/system/x11-common.service
-        - -lib/systemd/system/cryptdisks-early.service
-        - -lib/systemd/system/rcS.service
-        - -lib/systemd/system/cryptdisks.service
-    stage:
-      - $exclude-systemd-files
+      - -lib/systemd/system/cryptdisks-early.service
+      - -lib/systemd/system/cryptdisks.service
+      - -lib/systemd/system/hwclock.service
+      - -lib/systemd/system/rc.service
+      - -lib/systemd/system/rcS.service
+      - -lib/systemd/system/x11-common.service
+      - -tests
+
+  patchelf:
+    after:
+      - hotsos
+    plugin: nil
+    build-packages:
+      - file
+      - patchelf
+    override-prime: |
+      # patchelf
+      #
+      # To allow for a classic snap that works across multiple operating system
+      # runtime environments, we need to ensure all shipped binaries consume
+      # libraries from the core or shipped snap.  We accomplish this by patching
+      # `RPATH`` or interpreter into dynamically linked binaries.
+      #
+      # /snap/${CORE}/current/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+      # /snap/${CORE}/current/lib/aarch64-linux-gnu/ld-linux-aarch64.so.1
+      # /snap/${CORE}/current/lib/powerpc64le-linux-gnu/ld64.so.2
+      # /snap/${CORE}/current/lib/s390x-linux-gnu/ld64.so.1
+      CORE=core20
```

Comment:
TBH, this seemed a bit hacky at first glance to me, but even the [official blog post](https://snapcraft.io/blog/the-new-classic-confinement-in-snaps-even-the-classics-need-a-change) is recommending this approach. If I am understanding this correctly, the snap's `--classic` confinement causes application binaries to rely on system libraries instead of the shipped snap libraries. The default `--strict` confinement avoids this, but it's not suitable for hotsos's needs. I think `snapd` needs a middle ground between these two: `Use snap's root while dynamic loader loads the application dependencies, switch to system's root afterwards`. But this strategy may have some shortcomings if the application is loading its libraries at runtime (e.g. plugins loaded on-demand).

Reply:
I am not sure "rely on" is the right way of thinking about this. In a strictly
confined snap the filesystems are mounted into the snap namespace so that the
snap cannot "see" the host's filesystem. In a classically confined snap that's
not the case and the snaps are in `/snap/${SNAPNAME}/current`. This means that
if you, e.g., execute a shell script from your snap, the `bash` interpreter will
be the one from the host. The same is true for Python scripts. This behavior
might lead to unexpected consequences but will work most of the time.

In this case here the problem lies with what libraries the dynamic linker loads
when `hotsos` loads the `cryptography` module. There is a rust extension that
needs to be loaded, hence the involvement of the dynamic linker. At some point
in the library dependency chain, the loader will want to load `libphread.so` and
it will load the one from the host because of the aforementioned mounting
behavior of classic snaps. This causes a failure if the host is running a newer
version of `libc` than the one the rust extension in the snap was compiled with.
Using the hacky approach and patch the loader paths to include the
`/snap/${SNAPNAME}/....` prefix addresses this issue. Note that this also
requires that we install `libc` in the snap which is normally provided by the
`core??` snap itself. But since we can't patch anything in another snap we need
to install `libc` here ourselves.

I hope this clarifies the reason for patching elf files.